src/librbd/cache/pwl/rwl/WriteLog.cc

```diff
@@ -339,14 +335,12 @@ bool WriteLog<I>::initialize_pool(Context *on_finish, pwl::DeferredContexts &lat
       lderr(cct) << "Pool layout version is "
                  << D_RO(pool_root)->header.layout_version
                  << " expected " << RWL_LAYOUT_VERSION << dendl;
-      on_finish->complete(-EINVAL);
-      return false;
+      goto error_handle;
     }
     if (D_RO(pool_root)->block_size != MIN_WRITE_ALLOC_SIZE) {
       lderr(cct) << "Pool block size is " << D_RO(pool_root)->block_size
                  << " expected " << MIN_WRITE_ALLOC_SIZE << dendl;
-      on_finish->complete(-EINVAL);
-      return false;
+      goto error_handle;
```

Comment:
@CongMinYin what do you think about encapsulating this in a function like `handle_close`  https://github.com/ceph/ceph-ci/blob/92d7bc2869a92c18eda336e8aad551a77d26d08c/src/librbd/ImageState.cc#L876-L883

Reply:
The two situations are a little different. The example is to process all the results in the callback function, and the code is concise. Here is only one kind of result processing. There are four kinds of results, and just what I changed now need to call `pmemobj_close(m_log_pool);`, the sucessful one even not in this func:
```
on_finish->complete(-errno);
on_finish->complete(-pmemobj_tx_errno());
on_finish->complete(-EINVAL);
m_image_ctx.op_work_queue->queue(on_finish, 0);  // sucess, not in this func
```
If I just change this kind of error handle to func:

```
void handle_error(PMEMobjpool *m_log_pool, Context *on_finish) {
    pmemobj_close(m_log_pool);
    on_finish->complete(-EINVAL);
}
initialize_pool() {
    ...
   if (...) {
        handle_error(m_log_pool, on_finish);
        return false;
    }
}
```
It doesn't seem to be getting simpler. If you don't like `goto` statements, may leave it as it is and add `pmemobj_close(m_log_pool) `before each `onfinish`. TODO: Make the same changes to SSD mode.